### PR TITLE
add locks for libuv

### PR DESCRIPTION
**Pull request by JeffBezanson**
Rebased and squashed version of #29706.

### CODE REVIEW

File: src/init.c [hunk 1]
```diff
@@ -406,23 +410,31 @@ static void *init_stdio_handle(const char *stdio, uv_os_fd_t fd, int readable)
         break;
     case UV_NAMED_PIPE:
         handle = malloc(sizeof(uv_pipe_t));
+        JL_UV_LOCK();
         if ((err = uv_pipe_init(jl_io_loop, (uv_pipe_t*)handle, 0))) {
+            // JL_UV_UNLOCK() equivalent is done during unwinding
```
Comment (yuyichao) [inline src/init.c]: Errr, no. This is not the case for the NOGC version of the lock.

File: src/julia_internal.h [hunk 2]
```diff
@@ -108,6 +109,15 @@ static inline void jl_assume_(int cond)
 #  define JL_USE_IFUNC 0
 #endif
 
+// If this is detected in a backtrace of segfault, it means the functions
+// that use this value must be reworked into their async form with cb arg
+// provided and with JL_UV_LOCK used around the calls
+static uv_loop_t *const unused_uv_loop_arg = (uv_loop_t *)0xBAD10;
+
+extern jl_mutex_t jl_uv_mutex;
+#define JL_UV_LOCK() JL_LOCK_NOGC(&jl_uv_mutex)
```
Comment (yuyichao) [inline src/julia_internal.h]: Has this been tested under heavy GC to make sure that none of the locked code ever touches the GC (or better, by adding an assertion on GC entry that this lock is not held)?

Comment (vtjnash) [reply src/julia_internal.h]: No, we know for a fact that we hold this lock while doing arbitrary user computation

File: src/init.c (hunk 1 above)
Comment (JeffBezanson) [reply src/init.c]: I think the locks here can just be removed; threads haven't been started yet at this point.

**Pull request merged**